java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/VectorValueComparator.java

```diff
@@ -76,6 +88,10 @@ public void attachVector(V vector) {
   public void attachVectors(V vector1, V vector2) {
     this.vector1 = vector1;
     this.vector2 = vector2;
+
+    final boolean v1MayHaveNulls = vector1.getField().getFieldType().isNullable();
+    final boolean v2MayHaveNulls = vector2.getField().getFieldType().isNullable();
```

Comment:
For the cases I have tested the java vectors seem to have a validity buffer even when there can be no nulls. Depending on the `AllocationManager` this is some form of buffer called `EMPTY` which is not null but cannot allocate memory for items. It's not possible the test if the buffer is equal to one of these because they are `private`

However, given that someone might change the implementation one day I will include a check for a `null` validity buffer to catch this case as well.

Reply:
Ah, that's what I wasn't remembering. In that case, the right check might be to see if the buffer length is 0 (in which case there can't be any nulls)?

Reply:
Weirdly it seems that the validity buffer does have a capacity set when any values are added to the value buffer. There is a method to find the number of nulls in the vector which is relatively fast, only needs to be checked when attaching the vector and can be skipped for `FieldTypes` that are `nonNullable`

```
private boolean mayHaveNulls(V v) {
    return  v.getField().getFieldType().isNullable() ||
            v.getNullCount() > 0;
  }
```

Reply:
Ah, thanks for working through it. I suppose Java doesn't optimize the validity buffer in the same way that C++ does.